mm2src/coins/eth/eth_swap_v2/eth_taker_swap_v2.rs

```diff
@@ -457,37 +458,113 @@ impl EthCoin {
-                    &TAKER_SWAP_V2,
-                    EthPaymentType::TakerPayments,
-                    TAKER_PAYMENT_STATE_INDEX,
-                )
-                .await?;
-            if taker_status == U256::from(TakerPaymentStateV2::MakerSpent as u8) {
-                return Ok(taker_payment.clone());
-            }
             let now = now_sec();
             if now > wait_until {
-                return MmError::err(WaitForPaymentSpendError::Timeout { wait_until, now });
+                return MmError::err(FindPaymentSpendError::Timeout { wait_until, now });
             }
-            Timer::sleep(10.).await;
+
+            // Skip retrieving events if tx_hash is already found
+            if tx_hash.is_none() {
+                // get all logged TakerPaymentSpent events from `from_block` till current block
+                let events = match self
+                    .events_from_block(taker_swap_v2_contract, "TakerPaymentSpent", from_block)
+                    .await
+                {
+                    Ok(events) => events,
+                    Err(e) => {
+                        error!(
+                            "Error getting TakerPaymentSpent events from {} block: {}",
+                            from_block, e
+                        );
+                        Timer::sleep(5.).await;
+                        continue;
+                    },
+                };
+
+                if events.is_empty() {
+                    info!("No events found yet from block {}", from_block);
+                    Timer::sleep(5.).await;
+                    continue;
+                }
+
+                // this is how spent event looks like in EtomicSwapTakerV2: event TakerPaymentSpent(bytes32 id, bytes32 secret)
+                let found_event = events.into_iter().find(|event| &event.data.0[..32] == id.as_slice());
+
+                if let Some(event) = found_event {
+                    if let Some(hash) = event.transaction_hash {
+                        // Store tx_hash to skip fetching events in the next iteration if "eth_getTransactionByHash" is unsuccessful
+                        tx_hash = Some(hash);
+                    }
+                }
+            }
+
+            // Proceed to check transaction if we have a tx_hash
+            if let Some(tx_hash) = tx_hash {
+                match self.transaction(TransactionId::Hash(tx_hash)).await {
+                    Ok(Some(t)) => {
+                        let transaction = signed_tx_from_web3_tx(t).map_err(FindPaymentSpendError::Internal)?;
+                        return Ok(transaction);
+                    },
+                    Ok(None) => info!("spendTakerPayment transaction {} not found yet", tx_hash),
+                    Err(e) => error!("Get tx {} error: {}", tx_hash, e),
+                };
+                Timer::sleep(check_every).await;
+                continue;
+            }
+
+            Timer::sleep(5.).await;
```

Comment:
Two Qs
Q1: Why not use `check_every` everywhere? 
Q2: why not break and return if `self.transaction fails` ? since the only possible error `self.transaction(TransactionId::Hash(tx_hash)).await` could return is if serde fails and if we keep getting serde error for a `tx_hash` then this would become an infinite loop 

proposed loop

```rust
loop {
            let now = now_sec();
            if now > wait_until {
                return MmError::err(FindPaymentSpendError::Timeout { wait_until, now });
            }

            // get all logged TakerPaymentSpent events from `from_block` till current block
            let events = match self
                .events_from_block(taker_swap_v2_contract, "TakerPaymentSpent", from_block)
                .await
            {
                Ok(events) => {
                    if events.is_empty() {
                        info!("No events found yet from block {}", from_block);
                        Timer::sleep(check_every).await;
                        continue;
                    }
                    events
                },
                Err(e) => {
                    error!(
                        "Error getting TakerPaymentSpent events from {} block: {}",
                        from_block, e
                    );
                    Timer::sleep(check_every).await;
                    continue;
                },
            };

            let tx_hash = events.into_iter().find_map(|event| {
                // this is how spent event looks like in EtomicSwapTakerV2: event TakerPaymentSpent(bytes32 id, bytes32 secret)
                if &event.data.0[..32] == id.as_slice() {
                    event.transaction_hash
                } else {
                    None
                }
            });

            // Proceed to check transaction if we have a tx_hash
            if let Some(tx_hash) = tx_hash {
                match self.transaction(TransactionId::Hash(tx_hash)).await {
                    Ok(Some(t)) => {
                        let transaction = signed_tx_from_web3_tx(t).map_err(FindPaymentSpendError::Internal)?;
                        return Ok(transaction);
                    },
                    Ok(None) => {
                        info!("spendTakerPayment transaction {} not found yet", tx_hash);
                        Timer::sleep(check_every).await;
                        continue;
                    },
                    Err(e) => {
                        let error = format!("Get tx {} error: {}", tx_hash, e);
                        error!("{error}");
                        // TODO: return appropriate error
                    },
                }
            }

            Timer::sleep(check_every).await;
        }
```

Reply:
and lastly, we probably want to add limit to this loop

Reply:
Ok I missed this part of the fn ` if now > wait_until {
                return MmError::err(FindPaymentSpendError::Timeout { wait_until, now });
            }`

Reply:
Please re read this comment https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261#discussion_r1838047167 
```
But if we got None or error in next step after Some(tx_hash), we dont have to repeat the whole process from the beginning.
```

---
> Q1: Why not use check_every everywhere?

lets use it https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261/commits/f437dde06e77774407afc71829c29961faa3366a

> Q2
> and lastly, we probably want to add limit to this loop

https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261#discussion_r1842325488

Loop will be stoped once time is out